### ISSUE TITLE
ORDER BY with WHEREs performance hit

### ISSUE THREAD

**Issue by bhishp** [typeorm/typeorm#6813]
**Issue type:**

[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[x] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[x] `0.2.26`

**Steps to reproduce or a small repository showing the problem:**

Hi, we seem to be experiencing a big performance hit when using ORDER BY along with WHERE clauses in our queries vs native query performance. Tedious (mssql driver) regularly throws TimeoutErrors so I have done some investigation as per below.

**Background**

The issues are seen with a **View** that is defined within the database and mapped in TypeORM using the `@ViewEntity` decorator. Stub below:

```ts
@ViewEntity({
  name: 'BASELINE_VIEW',
  schema: CNS_CPO_MID_CAPACITY_SCHEMA,
})
export class BaselineView  {
  @ViewColumn()
  public HRecordID!: string;

  @ViewColumn()
  public BaselineID!: string;
  ....
}
```

The view does joins across many tables under-the-hood so there are no joins in my queryBuilder code. The users typically filters on lots of different criteria across several columns, therefore we may need to add a number of where clauses to the view query.

**Benchmarking results**

The benchmarking was done to compare TypeORM queryBuilder `getMany` performance vs tedious `execSql` performance when selecting **150 records only**. Tests were done in different combinations, specifically:

* Select * with Order By
* Select * with several Where clauses
* Select * with Order By and several Where clauses

| Method | Avg query time (seconds) |
| -------- | ----------------------- |
| **Tedious** | --- |
| Tedious with Order By | 1.384s |
| Tedious with several Where clauses | 1.540s |
| Tedious with Order By and several Where clauses | 5.430s |
| **TypeORM** | --- |
| TypeORM with Order By | 1.385s |
| TypeORM with several Where clauses | 2.462s |
| TypeORM with Order By and several Where clauses | **9.330s** |

As you can see, TypeORM stays pretty consistent with tedious in terms of performance in most cases, but once several Where clauses are used alongside Order By there seems to be a massive hit. I am assuming it must be something to do with TypeORM internals (beyond my knowledge) because the performance is fine when using tedious itself?

**Benchmarking Code**

Disclaimer: Benchmarks were carried out on my local machine against a deployed test database, which may have other devs interacting with at the same time, I ran each test 75 times to try and get a fair average in this circumstance.

Please feel free to critique my test code and let me know you find any issues with the benchmarking process itself.

The benchmarking was done in a quick n dirty script and executed in ts-node-dev locally. I created a queryBuilder against the view and then tweaked the queryBuilder definition for each test case. I printed the SQL generated by the query builder and used this generated SQL to drive the tedious query.

Note: Some column names and values have been garbled as this is closed-source code.

```ts
const ormConfig: SqlServerConnectionOptions = {
  type: process.env.DBTYPE as 'mssql',
  host: process.env.DBHOST,
  port: 1433,
  username: process.env.DBUSERNAME,
  password: process.env.DBPASS,
  database: process.env.DBDATABASE,
  schema: DB_SCHEMAS,
  synchronize: false,
  requestTimeout: 500000,
  options: {
    encrypt: true,
    packetSize: 16384,
  },
  entities: [entitiesPath],
  logging: process.env.DBLOGGING as LoggerOptions,
  logger: new TypeORMLogger(),
};

const runTypeORMTests = async () => {
  console.log('Creating connection');
  const connection = await createConnection(ormConfig);

  const typeORMQuery = getRepository(BaselineView)
    .createQueryBuilder()
    .andWhere(
      `(
        [F] = :F0 OR
        [F] = :F1 OR
        [F] = :F2
      )`,
      {
        F0: 'C',
        F1: 'LE',
        F2: 'MD',
      },
    )
    .andWhere(
      `(
        [G] = :G0 OR
        [G] = :G1
      )`,
      {
        G0: 'G/N',
        G1: '',
      },
    )
    .andWhere(
      `(
        [LR] = :LR0 OR
        [LR] = :LR1 OR
        [LR] = :LR2 OR
        [LR] = :LR3 OR
        [LR] = :LR4 OR
        [LR] = :LR5 OR
        [LR] = :LR6 OR
        [LR] = :LR7 OR
        [LR] = :LR8 OR
        [LR] = :LR9 OR
        [LR] = :LR10
      )`,
      {
        LR0: 'US',
        LR1: '',
        LR2: 'NW',
        LR3: 'WAF',
        LR4: 'PG',
        LR5: 'E',
        LR6: 'US WC',
        LR7: 'M',
        LR8: 'L',
        LR9: 'FSU+',
        LR10: 'Placeholder',
      },
    )
    .andWhere(
      `(
        [DR] = :DR0 OR
        [DR] = :DR1 OR
        [DR] = :DR2 OR
        [DR] = :DR3 OR
        [DR] = :DR4 OR
        [DR] = :DR5 OR
        [DR] = :DR6 OR
        [DR] = :DR7 OR
        [DR] = :DR8 OR
        [DR] = :DR9
    )`,
      {
        DR0: 'C',
        DR1: 'NW',
        DR2: 'M',
        DR3: 'L',
        DR4: 'WAF',
        DR5: 'FSU+',
        DR6: 'Placeholder',
        DR7: 'PG',
        DR8: 'E',
        DR9: 'US WC',
      },
    )
    .andWhere(`AB = :AB0`, { AB0: 'Y' })
    .andWhere(`([VTK] = :VTK0)`, { VTK0: 'B' })
    .addOrderBy(`[SourceDate]`, 'DESC')
    .addOrderBy(`[HRecordID]`, 'ASC')
    .skip(0)
    .take(150);

  const test = async (i: number) => {
    console.log('Query #: ', i);
    const endtimer = useTimer();
     await typeORMQuery.getMany();
    const hrend = endtimer();
    return hrend;
  };
  const benchmarkResults = await executePromiseSequence(typeORMTestCount, test);
  fs.writeFileSync('./typeorm-results.json', JSON.stringify(benchmarkResults, null, 2));
  connection.close();
};

const runTediousTests = async () => {
  const connection = new Connection(config);

  const query = `
  SELECT "BaselineView"."HRecordID" AS "BaselineView_HRecordID",
   "BaselineView"."BaselineID" AS "BaselineView_BaselineID",
   ...
  FROM "CNS_CPO_MID_CAPACITY"."BASELINE_VIEW" "BaselineView"
  WHERE (
      [F] = 'C' OR
      [F] = 'LE' OR
      [F] = 'MD'
  ) AND (
  -- WHERE (
      [G] = 'G/N' OR
      [G] = ''
  ) AND (
      [LR] = 'US' OR
      [LR] = '' OR
      [LR] = 'NW' OR
      [LR] = 'WAF' OR
      [LR] = 'PG' OR
      [LR] = 'E' OR
      [LR] = 'US WC' OR
      [LR] = 'M' OR
      [LR] = 'L' OR
      [LR] = 'FSU+' OR
      [LR] = 'Placeholder'
  ) AND (
      [DR] = 'C' OR
      [DR] = 'NW' OR
      [DR] = 'M' OR
      [DR] = 'L' OR
      [DR] = 'WAF' OR
      [DR] = 'FSU+' OR
      [DR] = 'Placeholder' OR
      [DR] = 'PG' OR
      [DR] = 'E' OR
      [DR] = 'US WC'
  ) AND
      [AB] = 'Y'
    AND
      [VTK] = 'B'
  ORDER BY [SourceDate] DESC, [HRecordID] ASC
  -- ORDER BY (SELECT NULL) -- <= used testing without ORDER BY
  OFFSET 0 ROWS FETCH NEXT 150 ROWS ONLY
  `;

  connection.on('connect', async (err: Error) => {
    if (err) {
      console.error(err);
      return;
    }

    const results: [[number, number]] = await Array.from(Array(tediousTestCount)).reduce(
      async (accumulator, promise, i) => {
        const acc = await accumulator;
        const res = await new Promise((res, rej) => {
          console.log('Query #: ', i);
          const hrstart = process.hrtime();
          const req = new t.Request(query, (error, rowCount) => {
            if (error) {
              rej(error);
              return;
            }
            const hrend = process.hrtime(hrstart);
            res(hrend);
          });
          connection.execSql(req);
        });
        acc.push(res);
        return acc;
      },
      Promise.resolve([]),
    );
    fs.writeFileSync('./tedious-results.json', JSON.stringify(results, null, 2));
    connection.close();
  });
};

(async () => {
  await runTypeORMTests();
  await runTediousTests();
})();
```

**Related issue?**

I have seen another ticket regarding performance here (https://github.com/typeorm/typeorm/issues/2381) but I feel this is different because we are not defining any joins within TypeORM itself - though there are several joins defined in the view.

**Question**

Is anyone aware of what could be causing the slowness to occur within the TypeORM query in this circumstance vs Tedious lib? Is anyone able to point me in the correct direction of how to dig deeper into the TypeORM internals that may be adding the overhead?

Thanks for your help :-)

**Comment by imnotjames** [comment]
Apologies if I missed it - did you retrieve the query generated by TypeORM & try running that natively?  Perhaps with an `EXPLAIN`?

**Comment by bhishp** [comment]
Thanks for your reply.

Yes, the query was generated by calling `query.getSql()` and the benchmark results are a comparison of TypeORM execution time vs native (tedious lib) execution time. I have done an explain against the query natively but can that be done within TypeORM too? The issue here is specifically comparing the native query time vs TypeORM. I assume there is some overhead to using the lib but jumping from ~5.4s to ~9.3s seems like a lot to me.

I realise my issue is a bit wordy so please ask any more questions to help get out the right details.

**Comment by bhishp** [comment]
Interestingly, the performance hit seems to be exponential. We have tested the same benchmarks in another Database instance with more data in and see the performances degrade as such:


_ | _
-- | --
Tedious with Order By and several Where clauses | 65.074s
TypeORM with Order By and several Where clauses | 435.193s

So almost 6.7x slower in this case
